src/metabase/query_processor/middleware/wrap_value_literals.clj

```diff
@@ -101,19 +104,129 @@
   [this info & _]
   [:absolute-datetime this (get info :unit :default)])
 
+(defmulti ^:private coerce-temporal
+  "Coerce temporal value `t` to `target-class`, or throw an Exception if it is an invalid conversion."
+  {:arglists '([t target-class])}
+  (fn [t target-class]
+    [(class t) target-class]))
+
+(defn- throw-invalid-conversion [message]
+  (throw (ex-info message {:type qp.error-type/invalid-query})))
+
+(defn- throw-invalid-date []
+  (throw-invalid-conversion (i18n/tru "Invalid date literal: expected a date, got a time")))
+
+(defmethod coerce-temporal [java.time.LocalDate      java.time.LocalDate] [t _target-class]  t)
+(defmethod coerce-temporal [java.time.LocalTime      java.time.LocalDate] [_t _target-class] (throw-invalid-date))
+(defmethod coerce-temporal [java.time.OffsetTime     java.time.LocalDate] [_t _target-class] (throw-invalid-date))
+(defmethod coerce-temporal [java.time.LocalDateTime  java.time.LocalDate] [t _target-class]  (t/local-date t))
+(defmethod coerce-temporal [java.time.OffsetDateTime java.time.LocalDate] [t _target-class]  (t/local-date t))
+(defmethod coerce-temporal [java.time.ZonedDateTime  java.time.LocalDate] [t _target-class]  (t/local-date t))
+
+(defn- throw-invalid-time []
+  (throw-invalid-conversion (i18n/tru "Invalid time literal: expected a time, got a date")))
+
+(defn- LocalTime->OffsetTime [t]
+  (if (= (qp.timezone/results-timezone-id) "UTC")
+    (t/offset-time t (t/zone-offset 0))
+    ;; if the zone is something else, we'll just have to make do with a LocalTime, since there's no way to determine
+    ;; what the appropriate offset to use for something like `US/Pacific` is for a give TIME with no DATE associated
+    ;; with it.
+    t))
+
+(defmethod coerce-temporal [LocalDate      LocalTime] [_t _target-class] (throw-invalid-time))
```

Comment:
I feel that a few comment blocks explaining some of the things to consider would be helpful.

- All Local to Offset/Zoned conversions must consider `results-timezone-id`
- All Offset/Zoned to Local conversions do not need to consider `results-timezone-id` because session timezone will already be applied (is that correct?)
- Others?